developing/on-call/README.md

```diff
@@ -0,0 +1,193 @@
+  * 5 minutes after the alert, notify via voice call.
+* Your on-call rotation should have an escalation policy that escalates
+  from the primary to secondary after no more than 10 minutes, and from
+  secondary to tertiary after no more than an additional 10 minutes.
+  Optimally, this should be as short as possible to ensure that there is
+  a quick response; remember that an alert going unnoticed can incur a
+  significant SLO impact. A 99.99% uptime requires no more than 13
+  minutes of downtime a quarter for instance -- if you have a 10 minute
+  escalation, an alert that falls through to the secondary may blow the
+  SLO on its own if the problem is serious enough.
+
+## Project Expectations
+
+In addition to the expectations we have for on-call engineers, there are
+also expectations we make for the project we are on-call for in order to
+ensure that on-call is not an undue burden.
+
+* The on-call rotation *should not be getting more than 2-3 alerts per
+  day*, and even that is bordering on excessive, especially if these are
+  off-hours. Optimally, this should be no more than 2-3 alerts *per shift*.
+  If the on-call burden for that rotation is higher than that, there
+  should be an understanding across engineering and product that engineering
+  time needs to be devoted to reducing the on-call burden. This could mean
+  relaxing SLOs or tuning alert thresholds, but it may also mean a deeper
+  investigation, bug fixing, or code and/or infra improvements to prevent
+  problems. The "SRE" way to do this is formal [SLOs](https://www.youtube.com/watch?v=tEylFyxbDLE)
+  and [error budgets](https://www.youtube.com/watch?v=y2ILKr8kCJU), but
+  they aren't always the right choice if the project is small or does
+  not have the constraints that come with a 24/7 web service.
+* Engineers who are primary or secondary on-call should essentially be
+  considered off project work; they should focus on taking care of
+  immediate needs like writing or tuning alerts, fixing stability-threatening
+  bugs, addressing reported security vulnerabilities, or updating
+  documentation. If they can contribute to project work as well, that
+  should be a bonus, not an expectation.
+* Alerts for any project should be well-documented so that an on-call
+  engineer can at least begin the process of diagnosis. Questions this
+  documentation should answer include:
+  * What does this alert mean, literally?
+  * What are common causes for this alert to fire?
+  * What logs, tools, or other resources can I use to find out more about
+    why this alert fired?
+* It's not unusual for us to have infra teams that are significantly
+  smaller than application development teams on a project. For this
+  reason, it is probably a good idea for all engineers to be involved
+  in any sort of infra rotation; however, you should make sure that any
+  single rotation has an infra engineer in the escalation path.
```

Comment:
We've just removed Infra as part of the escalation path on MilMove. The idea is that infra can be used as SME's for any issues but they'd have to be paged separately outside of work or SLA hours. 

I'd maybe like to turn this on its head and say that application teams should look at their code as a set of "systems" and own those systems including the infrastructure tools they need to understand health and debug the system.

Reply:
So, I definitely agree that application teams should understand the infrastructure that their applications are built on enough to have a good mental model of how the components work and at least make some initial attempts at diagnosing infrastructure issues. I think that is one of the reasons to have application engineers go through the infra on-call escalation regularly.

On the other hand, I'm wary of just yanking infra folks out of day-to-day on-call. It's good as both an empathy building tool, and in seeing how the application actually works on the infrastructure so that infra folks can make good recommendations about how to improve reliability and design new systems to support those applications.

Reply:
What I'm particularly afraid of here is saying "in the escalation path" as though an infra engineer ought to be able to solve any problem. This is what was happening on milmove: an alert related to the app would go through, it would have an affect that appeared to be infra related, it would get escalated, infra would have to figure out what was causing the problem only to determine it was the app being the problem, and then send it back.  The tendency is then to say anything that looks infra related (which can be as tenuous as "alert came from AWS") needs to go directly to Infra and skip the app teams. 

I'd prefer that infra and app devs be on different escalation paths (each having their own primary/secondary/tertiary) and use each other as SMEs but retain the ownership of the event while an investigation is ongoing.  I want to avoid the pager hot-potato that seems to happen on MilMove.

Reply:
Well, I think it's wise for at least one infra engineer to be in the *infra* escalation path. Putting them in the application engineering path makes less sense (although optimally, they should be getting some exposure to the application, and it would also force application engineers to document their alerts and responses).